sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +164,415 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
     result
   }
 }
+
+final class GenericIntArrayData(private val primitiveArray: Array[Int])
+  extends GenericArrayData(Array.empty) {
+  override def array(): Array[Any] = primitiveArray.toArray
+
+  override def copy(): ArrayData = new GenericIntArrayData(primitiveArray)
+
+  override def numElements(): Int = primitiveArray.length
+
+  override def isNullAt(ordinal: Int): Boolean = false
+  override def getInt(ordinal: Int): Int = primitiveArray(ordinal)
+  override def toIntArray(): Array[Int] = {
+    val array = new Array[Int](numElements)
+    System.arraycopy(primitiveArray, 0, array, 0, numElements)
+    array
+  }
+  override def toString(): String = primitiveArray.mkString("[", ",", "]")
+
+  override def equals(o: Any): Boolean = {
```

Comment:
At what level do we define equality? Should we be able to compare an Unsafe row to a safe one, and get the same results if contain the same data? If so, this equals method will no suffice. It will also mean that equals is going to be pretty expensive. If not, please take a look at the java.util.Arrays class; it implements much of the code you are writing here.

Same goes for hashcode...